### PR TITLE
Replace GC in std.string.abbrev with malloc/free

### PR DESCRIPTION
**Pull request by WalterBright**
http://dlang.org/phobos/std_string.html#.abbrev

It still allocates for the returned associative array, but that can't be helped at the moment. At least it doesn't leak now.

### CODE REVIEW

File: std/string.d [hunk 1]
```diff
@@ -4789,17 +4798,45 @@ body
...
-
-    // Make a copy when sorting so we follow COW principles.
-    values = values.dup;
-    sort(values);
+    S[S] result;
 
+    // Make a copy so it can be sorted in place
     size_t values_length = values.length;
+
+    S* vptr;
```
Comment (JakobOvrum) [inline std/string.d]: Just use `vsorted.ptr`?

Comment (WalterBright) [inline std/string.d]: that would work, but would make the code harder to follow

File: std/string.d [hunk 2]
```diff
@@ -4789,17 +4798,45 @@ body
...
+    S[] vsorted;
+    if (__ctfe)
+        vsorted = values.dup;
+    else
+    {
+        void alloc() @trusted nothrow @nogc
```
Comment (JakobOvrum) [inline std/string.d]: This could be replaced by #3031 if/when that is merged (just cross-referencing so I can do the work then).

**Comment by JakobOvrum** [comment]
This is a breaking change as `abbrev` is no longer `pure`. Should we wait for https://github.com/D-Programming-Language/druntime/pull/1183 ?

**Comment by monarchdodra** [comment]
> At least it doesn't leak now.

What does that mean? Using the GC is "leak" now?

IMO, this pull is WAY overkill in terms of amounts of code added for the result. In particular, I object to such manual memory allocation for such a basic result. This kind of code does not belong here, and should be encapsulated in a library that provides ways to create Non-GC arrays with scoped lifetimes. IMO, "UniqueArray" could be a good starting point.

Either that, or wait until we have a working implementation for ref counted arrays, and then use that.

**Comment by JakobOvrum** [comment]
It would be less code if #3031 would be applied, but I agree `Unique[Array]` and/or `RefCounted[Slice]` is the way to go for these kinds of applications. (#3031 is intended to guarantee memory safety in generic code, not provide convenience)

However, that doesn't mean a patch that works should be rejected on those grounds. Unique references and reference counting are much harder tasks and so can't be expected in the short-term. If someone is willing to hot-patch issues in the meantime, we should accept them on the simple grounds that they do improve the state of affairs for users as long as someone is willing to thoroughly review them.

AFAICS, the question is simply whether or not this patch provides any value, and I think the answer is yes, because it does provide `@nogc`. I'm worried about the fact that it simutaneously introduces impurity.

**Comment by monarchdodra** [comment]
> I think the answer is yes, because it does provide `@nogc`.

My point was that it *doesn't*, because it still returns an AA. All this does is get rid of a single GC allocation, and that's it.

The unittest is now not marked as pure, but it wasn't marked as `@nogc`

**Comment by JakobOvrum** [comment]
> My point was that it doesn't, because it still returns an AA. All this does is get rid of a single GC allocation, and that's it.

You're right, my mistake. I still see the value in eliminating short-lived GC allocations, but I definitely don't think it's worth sacrificing `pure` for it.

As for reference counting, we do have a reference-counted, sortable array type in Phobos: std.container.array.Array. It has the same inherent problem with purity, though.

**Comment by kuettler** [comment]
As an unknowledgeable observer: Why does this function exist in the first place? I understand it is there and will not go away. Still, the functions seems rather specific. I would not expect this function in a standard library.

**Comment by WalterBright** [comment]
It's there because it exists in ruby/python.

**Comment by kuettler** [comment]
It seems such a function exists in ruby. I do not know of a python version.

True to spirit there should be a `abbreviater` that transforms a SortedRange to a range of tuples. If indeed the functionality is needed. Otherwise, this function might be better of in `std.attic.string`.

File: std/string.d [hunk 3]
```diff
@@ -4763,6 +4763,15 @@ body
  * in one of a known set of strings, and the program will helpfully
  * autocomplete the string once sufficient characters have been
  * entered that uniquely identify it.
+ *
+ * Params:
+ *      array of strings
```
Comment (andralex) [inline std/string.d]: I appreciate the effort of adding Params: and Returns:, but just mentioning the types is hardly helpful.

File: std/string.d [hunk 4]
```diff
@@ -4789,17 +4798,45 @@ body
...
+    S* vptr;
+    S[] vsorted;
+    if (__ctfe)
+        vsorted = values.dup;
+    else
+    {
+        void alloc() @trusted nothrow @nogc
+        {
+            import core.exception : onOutOfMemoryError;
+            vptr = cast(S*)malloc(values_length * (S).sizeof);
```
Comment (andralex) [inline std/string.d]: why the paren around S

**Comment by andralex** [comment]
>This change has no value.

Not sure how you meant it but if you went "eye for an eye" it would be nice to avoid that.

**Comment by andralex** [comment]
@WalterBright: this function has been mehworthy to begin with. The way I see this is either introduce this improvement as a 100% backward-compatible change (i.e. keep it pure), or create a new function entirely. Right now it's sort of in between: it breaks compatibility and eliminates some but not all garbage (still returns an associative array). @Dicebot's idea of using a callback sounds like a nice angle for future work.

File: std/string.d [hunk 5]
```diff
@@ -4808,13 +4845,13 @@ string[string] abbrev(string[] values) @safe pure
...
-        string value = values[i];
+        string value = vsorted[i];
 
         // Skip dups
         for (nexti = i + 1; nexti < values_length; nexti++)
         {
-            nv = values[nexti];
-            if (value != values[nexti])
+            nv = vsorted[nexti];
+            if (value != vsorted[nexti])
```
Comment (schveiguy) [inline std/string.d]: Why not `if (value != nv)`

**Comment by schveiguy** [comment]
I find the usage of memory with an AA kind of wasteful here. Why not just return a type that is backed by a sorted array and use binary search to get the matching range? Seems it would be more useful to me.

File: std/string.d (hunk 3 above)
Comment (WalterBright) [inline std/string.d]: improved

File: std/string.d (hunk 4 above)
Comment (WalterBright) [inline std/string.d]: fixed

File: std/string.d (hunk 5 above)
Comment (WalterBright) [inline std/string.d]: changed

**Comment by WalterBright** [comment]
@Dicebot the purpose of @nogc is not "no allocations", it is to address the oft-stated desire of many programmers to not have allocations on the GC heap so there won't be an accumulation of garbage and an eventual collection pause. Replacing GC allocations with malloc/free is entirely consistent with that goal.

**Comment by WalterBright** [comment]
> Why not just return a type that is backed by a sorted array and use binary search to get the matching range?

That's exactly the sort of application that hash tables are better at, especially as the size of the data to be searched increases.

**Comment by WalterBright** [comment]
@andralex the purity issue will be resolved when we have a way to have a malloc/free pair in pure code. There's no reason we can't do that. https://github.com/D-Programming-Language/phobos/pull/3031 might be the solution.

**Comment by schveiguy** [comment]
> That's exactly the sort of application that hash tables are better at, especially as the size of the data to be searched increases

I mean, for example, I have a list of words:

forth, fortnight, fourth

As I type:
f => everything
fo => everything
for => forth, fortnight
fort => forth, fortnight
forth => forth

As the comment for the function says, an auto-complete box is exactly the application. I don't see how a hash table can give you a list of possible matches, you are only returning one possible match per try. In some cases, you don't produce any matches:

f => null
fo => null
for => null
fort => null
forth => forth

This seems like horrid behavior for a auto-complete box. The function I envision supplants the above, because you get the same information by checking the size of the result and/or looking at the first element.

Can't say I'd ever see using this function, it seems to have weird quirks and not much usefulness.

**Comment by WalterBright** [comment]
> I don't see how a hash table can give you a list of possible matches

```
foreach (key, value; aa)
```

> In some cases, you don't produce any matches:

Meaning those don't have unique abbreviations.

**Comment by WalterBright** [comment]
> Phobos shouldn't do any allocations at all on his own and here you just hide the problem. For example, on microcontrollers you are likely to not have  malloc  available either.

If running on a system without malloc, one has special problems even with C, as several functions in the C Standard library call malloc. Finding this is not too hard, as the program won't link if there's no malloc in the library. @nogc, and what it guarantees, is exactly what people have been asking for for many years.

That said, I think you'll agree that the recent Phobos PRs I've submitted have been aimed at:

1. eliminating all allocation
2. failing that, minimizing allocation
3. failing that, using malloc/free where such is entirely self-contained

**Comment by WalterBright** [comment]
Changing the api of this function is further than I'm willing to go at the moment. It's not an important function, and there is a lot more work in the rest of Phobos that's much more valuable.

> No, I was asking specifically for  @noheap  all this time

Unfortunately, that is unimplementable in a reliable fashion as a builtin attribute. The compiler can't reasonably know if it is calling malloc, because malloc can wear all kinds of costumes unknown to the compiler. This is not the case for the gc because the gc calls are hardwired into the compiler.

**Comment by MartinNowak** [comment]
Callback would be nice and rather trivial to implement.

**Comment by andralex** [comment]
@WalterBright the short of this is: keep this `@safe pure` or close this. Thanks!

**Comment by andralex** [comment]
@WalterBright ping - `@safe pure` or close please.

**Comment by DmitryOlshansky** [comment]
I guess time is up. No pure - no pull.

**Pull request closed without merging**